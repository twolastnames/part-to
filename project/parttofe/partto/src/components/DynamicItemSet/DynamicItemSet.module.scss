.dynamicItemSet {
  display: flex;
  width: 100%;
}

@mixin innerContent {
  padding: 16px;
}

.emptyPage {
  @include innerContent();
}
