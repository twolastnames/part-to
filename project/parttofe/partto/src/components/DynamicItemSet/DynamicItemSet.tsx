import React from "react";

import classes from "./DynamicItemSet.module.scss";
import { IconType } from "../Icon/Icon";
import { Empty } from "./Empty/Empty";
import { One } from "./One/One";
import { Multiple } from "./Multiple/Multiple";
import { ButtonSet } from "./ButtonSet/ButtonSet";

export interface Operation {
  text: string;
  icon: IconType;
  onClick: () => void;
}

export interface Item {
  listView: React.ReactElement;
  detailView: React.ReactElement;
  itemOperations: Array<Operation>;
}

export interface DynamicItemSetProps {
  items: Array<Item>;
  setOperations: Array<Operation>;
  emptyPage: React.ReactElement;
}

export function DynamicItemSet({
  setOperations,
  items,
  emptyPage,
}: DynamicItemSetProps) {
  return (
    <div className={classes.dynamicItemSet} data-testid="DynamicItemSet">
      <ButtonSet operations={setOperations} />
      {items.length === 0 ? (
        <Empty content={emptyPage} />
      ) : items.length === 1 ? (
        <One item={items[0]} />
      ) : (
        <Multiple items={items} />
      )}
    </div>
  );
}
