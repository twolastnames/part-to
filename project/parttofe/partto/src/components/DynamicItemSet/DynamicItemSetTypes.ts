import { ContextDescription } from "../../providers/DynamicItemSetPair";
import { IconType } from "../Icon/IconTypes";

export interface Operation {
  text: string;
  icon: IconType;
  onClick: () => void;
}

export interface Item {
  key: string;
  listView: React.ReactElement;
  detailView: React.ReactElement;
  itemOperations: Array<Operation>;
}

export interface DynamicItemSetProps {
  items: Array<Item>;
  setOperations: Array<Operation>;
  emptyPage: React.ReactElement;
  context: ContextDescription;
  pausedByDefault?: boolean;
}
