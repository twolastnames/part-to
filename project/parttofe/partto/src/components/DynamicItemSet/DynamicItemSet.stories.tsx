import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { DynamicItemSet } from "./DynamicItemSet";
import { ShellProvider } from "../../ShellProvider";
import { Next, Start } from "../Icon/Icon";

const meta: Meta<typeof DynamicItemSet> = {
  component: DynamicItemSet,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof DynamicItemSet>;

export const Empty: Story = {
  args: {
    items: [],
    setOperations: [
      {
        icon: Start,
        text: "start cooking",
        onClick: () => console.log("clicked"),
      },
      {
        icon: Next,
        text: "mark complete",
        onClick: () => console.log("clicked"),
      },
    ],
    emptyPage: <div>Empty Yo</div>,
  },
};

export const Single: Story = {
  args: {
    items: [
      {
        listView: <div>Should never see this</div>,
        detailView: <div>This is a detail view</div>,
        itemOperations: [
          {
            icon: Start,
            text: "start cooking",
            onClick: () => console.log("clicked"),
          },
          {
            icon: Next,
            text: "mark complete",
            onClick: () => console.log("clicked"),
          },
        ],
      },
    ],
    setOperations: [
      {
        icon: Start,
        text: "start cooking",
        onClick: () => console.log("clicked"),
      },
      {
        icon: Next,
        text: "mark complete",
        onClick: () => console.log("clicked"),
      },
    ],
    emptyPage: <div>Empty Yo</div>,
  },
};
