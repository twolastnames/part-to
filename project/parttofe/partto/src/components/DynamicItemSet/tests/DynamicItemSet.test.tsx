import React from "react";
import { describe, expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { DynamicItemSet } from "../DynamicItemSet";

describe("DynamicItemSet", () => {
  beforeAll(() => {});

  test("snapshot", () => {
    render(
      <DynamicItemSet
        items={[]}
        setOperations={[]}
        emptyPage={<div>still empty</div>}
      />,
    );
    const component = screen.getByTestId("DynamicItemSet");
    expect(component).toMatchSnapshot();
  });
});
