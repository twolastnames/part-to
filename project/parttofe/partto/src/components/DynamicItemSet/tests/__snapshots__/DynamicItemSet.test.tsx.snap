// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicItemSet snapshot 1`] = `
<div
  class="dynamicItemSet"
  data-testid="DynamicItemSet"
>
  <div
    class="buttonSet"
    data-testid="ButtonSet"
  />
  <div
    class="empty"
    data-testid="Empty"
  >
    <div>
      still empty
    </div>
  </div>
</div>
`;
